### PR TITLE
Fix Part of #4236 : Add tests for AsyncResultSubject

### LINKED ISSUE #4236: Add remaining tests for AsyncResultSubject.
### Is your feature request related to a problem? Please describe.

The class [AsyncResultSubject](https://github.com/oppia/oppia-android/blob/3ffeedf7e90795ae9b23ec600a19d7cff72f1ca4/testing/src/main/java/org/oppia/android/testing/data/AsyncResultSubject.kt#L28) (in `org.oppia.android.testing.data`) provides custom Truth subjects for verifying properties of `AsyncResult`. While a good number of tests are already written in [AsyncResultSubjectTest](https://github.com/oppia/oppia-android/blob/3ffeedf7e90795ae9b23ec600a19d7cff72f1ca4/testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt#L14), not all methods and edge cases are covered. This results in partial test coverage and missed validation of important behaviors.

### Describe the solution you'd like

We should expand `AsyncResultSubjectTest` to ensure **all public methods in `AsyncResultSubject` are thoroughly tested**, covering both success and failure paths. This will help guarantee correctness, improve reliability, and prevent regressions in future changes.

### Tasks / Checklist

Add new tests for the following scenarios:

1. **`isSuccessThat()`**: Already partially tested with strings, but add more tests for other types (e.g., nullable values, mismatched casts).
2. **`isComparableSuccessThat()`**: Verify with comparable types (e.g., `Int`, `String`) including correct ordering and type mismatch failures.
3. **`isProtoSuccessThat()`**: Add tests using a simple `MessageLite` proto to confirm correct proto assertions.
4. **`isFailureThat()`**: More validation on error chains, custom exceptions, and verifying root causes.
5.  **`isNewerOrSameAgeAs()` and `isOlderThan()`**: Add tests to validate correct age comparison between different `AsyncResult` instances.
6.  **`hasSameEffectiveValueAs()`**: Add more edge case tests (e.g., comparing pending vs failure, null vs non-null values).
7.  **Type mismatch scenarios**: Expand coverage to ensure incorrect type extraction (`extractSuccessValue`, `isIntSuccessThat`, etc.) throws the expected assertion error.

### Acceptance Criteria

* Each of the above methods should have **at least one dedicated test** verifying correct behavior.
* Negative test cases (incorrect types, mismatched expectations) should also be included.
* Ensure full branch and statement coverage for `AsyncResultSubject`.
* Follow existing test style in `AsyncResultSubjectTest` (Truth assertions, clear test names).

### PR Submission Directions

* Update `AsyncResultSubjectTest` with the new test cases.
* Run all tests with Robolectric to verify correctness.
* Ensure code passes `ktlint`, `bazel test`, and static analysis.

### PR DESCRIPTION
**Pull request by TanishMoral11**
## Explanation
 Fixes part of [#4236](https://github.com/oppia/oppia-android/issues/4236): Adds tests to ensure correct handling of non-iterable async results.

This PR adds tests for the `AsyncResultSubject` class. It verifies that the `AsyncResultSubject` handles non-iterable results correctly and ensures that the expected failures occur when the result is not iterable. The test specifically checks that the appropriate behavior is triggered when a non-iterable result is provided.

Updated Tests : 

Image

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes PTAL, and suggest me if there are any changes to make.

Image

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes,  PTAL !!

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes , could you please take a look ? 

Thanks !!

File: testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel [hunk 1]
```diff
@@ -33,3 +33,21 @@ oppia_android_test(
         "//utility/src/main/java/org/oppia/android/util/networking:debug_module",
     ],
 )
+
+# Adding test for AsyncResultSubject
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel]: This comment is not required.

File: testing/src/main/java/org/oppia/android/testing/data/AsyncResultSubject.kt
Comment (adhiamboperes) [inline testing/src/main/java/org/oppia/android/testing/data/AsyncResultSubject.kt:1]: remove this file from `test_file_exemptions`

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package org.oppia.android.testing.data
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.android.util.data.AsyncResult
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]:
```suggestion
@RunWith(JUnit4::class)
```

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+class AsyncResultSubjectTest {
+
+  @Test
+  fun testAsyncResultPending_isPending() {
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]:
```suggestion
  fun testAsyncResultSubject_pendingResult_checkIsPending() {
```

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    val pendingResult: AsyncResult<String> = AsyncResult.Pending()
+
+    AsyncResultSubject.assertThat(pendingResult).isPending()
+  }
+
+  @Test
+  fun testAsyncResultPending_isNotSuccess() {
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]:
```suggestion
  fun testAsyncResultSubject_pendingResult_checkIsNotSuccess() {
```

Comment (adhiamboperes) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: You are also missing a corresponding "isNotFailure()" test.

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+  @Test
+  fun testAsyncResultSuccess_hasSuccessValueWhere_matchesExpected() {
+    val successResult: AsyncResult<String> = AsyncResult.Success("Success value")
+
+    AsyncResultSubject.assertThat(successResult).hasSuccessValueWhere {
+      // Here we are verifying that the value is "Success value"
+      assertThat(this).isEqualTo("Success value")
+    }
+  }
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: Suggestion to reduce ambiguity.
```suggestion
  @Test
  fun testAsyncResultSuccess_hasSuccessValueWhere_matchesExpected() {
    val successResult: AsyncResult<String> = AsyncResult.Success("Some string")

    AsyncResultSubject.assertThat(successResult).hasSuccessValueWhere { result ->
      assertThat(result).isEqualTo("Some string")
    }
  }
```

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @TanishMoral11!
I noticed that you have not fully tested the class. Additionally, please rename the tests following the pattern that I have suggested.

File: testing/src/main/java/org/oppia/android/testing/data/AsyncResultSubject.kt
Comment (TanishMoral11) [reply testing/src/main/java/org/oppia/android/testing/data/AsyncResultSubject.kt:1]: Done .

File: testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel (hunk 1 above)
Comment (TanishMoral11) [reply testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel]: Done.

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt (hunk 3 above)
Comment (TanishMoral11) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt (hunk 2 above)
Comment (TanishMoral11) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: I tried switching to JUnit4 as suggested, but I encountered an error:  
`java.lang.UnsatisfiedLinkError: 'long android.os.SystemClock.uptimeMillis()'`.  
Since this is related to Android-specific functionality, I’m reverting to `RobolectricTestRunner` to ensure the tests run correctly.

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes , PTAL !!

Also , I believe I have now covered all the remaining tests.

Image
Thanks .

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes , PTAL !!
This PR is ready for review.
Thanks.

**Comment by TanishMoral11** [comment]
Hi @manas-yu  , could you please take a look ?
Thanks !!

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt (hunk 2 above)
Comment (adhiamboperes) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: That sounds good, thanks!

Comment (adhiamboperes) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: As per yesterday’s discussion, Ben mentioned that we don’t generally use the Robolectric runner, so one way to fix this error while using the JUnit runner is to inject FakeOppiaClock in this class, and then add an @before method to initialize the FakeOppiaClock object to uptimemillis or fake time mode as needed for these tests

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.android.util.data.AsyncResult
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+import org.robolectric.shadows.ShadowSystemClock
+
+@RunWith(RobolectricTestRunner::class)
+@Config(shadows = [ShadowSystemClock::class])
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: Use FakeOppiaClock as mentioned above.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
@TanishMoral11 PTAL at the remaining comments and the failing CI checks.

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt (hunk 6 above)
Comment (TanishMoral11) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: Done ..

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes , PTAL 
This pr is ready for review . also i attach updated tests image in pr description.
Thanks !!

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 7]
```diff
@@ -82,10 +85,49 @@ class AsyncResultSubjectTest {
   }
 
   @Test
-  fun testAsyncResultSubject_twoSuccessResults_checkNewerOrSameAge() {
-    val successResult1: AsyncResult<String> = AsyncResult.Success("First")
-    val successResult2: AsyncResult<String> = AsyncResult.Success("Second")
-    AsyncResultSubject.assertThat(successResult1).isNewerOrSameAgeAs(successResult2)
+  fun testAsyncResultSubject_newerResult_checkIsNewerOrSameAge() {
+    val olderResult: AsyncResult<String> = AsyncResult.Success("Older")
+    Thread.sleep(50)
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: This is not a typical Oppia pattern, please use  `testCoroutineDispatchers.advanceTimeBy(TimeUnit.SECONDS.toMillis(50))` instead.

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt [hunk 8]
```diff
@@ -5,9 +5,12 @@ import com.google.common.truth.Truth.assertThat
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.oppia.android.util.data.AsyncResult
+import org.robolectric.annotation.Config
+import java.io.FileNotFoundException
 
 /** Tests for [AsyncResultSubject]. */
 @RunWith(AndroidJUnit4::class)
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt:12]: Is there an error running with JUnit4?
```suggestion
@RunWith(JUnit4::class)
```

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
@TanishMoral11, some of the tests still don't meet the expected naming patters, e.g the tests that end in `_*That(). We want tests to read like a sentence e.g. `fun testAsyncResultSubject_conditionBeingTested_expectedOutcome()`. PTAL.

Also, I found it interested that the code coverage has decreased from 87% to 77%.

**Comment by TanishMoral11** [comment]
Hi @adhiamboperes , PTAL !!

I tried to run the test with JUnit4, but I am getting many errors like `java.lang.UnsatisfiedLinkError: 'long android.os.SystemClock.uptimeMillis()'`. I also tried the `Fake Oppia clock` method, but it’s not solving the issue.

Additionally, I’ve never seen JUnit4 and FakeOppiaClock used together in any of my test classes, so I’m unsure if Fake Oppia is the correct way to handle this.

Also, should I continue with AndroidJUnit4, as all tests are running as expected without any errors?

Thanks !!

File: testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt (hunk 7 above)
Comment (TanishMoral11) [reply testing/src/test/java/org/oppia/android/testing/data/AsyncResultSubjectTest.kt]: Done .

File: testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel [hunk 9]
```diff
@@ -41,6 +41,7 @@ oppia_android_test(
     test_class = "org.oppia.android.testing.data.AsyncResultSubjectTest",
     test_manifest = "//testing:test_manifest",
     deps = [
+        "//model/src/main/proto:languages_java_proto_lite",
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel]: Is this specifically required?

**Review by adhiamboperes — APPROVED** [review]
Thanks @TanishMoral11! It makes sense to use the AndroidJUnit4. 

I have re-run the failing coverage check, but if it doesn't pass again please investigate it. I have left a minot non-blocking comment.

File: testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel (hunk 9 above)
Comment (TanishMoral11) [reply testing/src/test/java/org/oppia/android/testing/data/BUILD.bazel]: Thanks @adhiamboperes , for pointing this out. I added it to test something else, but it is no longer required, so I am removing it now.

**Comment by TanishMoral11** [comment]
Hi @adhiamboperes , PTAL
All the CI checks have passed now.
Thanks !!
